### PR TITLE
#181414761 feature(user-profile):create user user profile

### PR DESCRIPTION
**Pull request by Procech20**
### What does this PR do?
Create user profile after registration.

### Description of Task to be completed?
- Create user profile route
- Tested the route
- Documented the route

### How should this be manually tested?
To test the changes manually:
* clone repo on your local machine
* switch to develop branch  by typing the command ```git checkout ft-user-profile-181414761```
* Run: ```yarn install```: to install all dependencies
* Create .env file
* Create a ```.env``` and the variables as mentioned in the ```.env.example``` file.
* Open the host address using an HTTP client such as ``postman``` or any other.
* Make ``GET`` or  ``PUT`` on  ``/api/v2/user/profile/:id`` to retrieve user's profile and/or update it.

### What to expect after running the commands:
* _On successful, A user will get a status of ``200`` for profile fetch and delete and ``201`` for profile update_

* _If the user doesn't exist, you'll a message that says: ``No such User was found`` with the status of: ``404``._

### Any background context you want to provide?
- N/A

``` Useful links ```:
- N/A

### What are the relevant pivotal tracker stories?
* https://www.pivotaltracker.com/story/show/181414761

### Screenshots (if appropriate)
- N/A

### Questions:
- N/A

---

### Task implementation plan

#1. Name of task
- Profile Page Settings

#2. Pivot Tracker ID

- [181414761](https://www.pivotaltracker.com/story/show/181414761)

#3. Branch Name

- ft-create-user-profile-181414761

#4. Description of task (Paraphrase what we have in Pivot Tracker)

* After signup on Barefoot Nomad, a user should be able to have their profile created and may update the account to add some more information.

#5. Acceptance criteria (Checks to be made)

- [ User should be able to Edit my profile, but should NOT be able to edit another user's profile]

```Scenario```: Given User registers to be a member of Barefoot Nomad.Registration is done.User Profile is created.A timestamp is associated to the profile

#6. Database schema (Migration & models)
 
Screen Shot 2022-04-18 at 18 40 37

#7. Validations point (Input to validate)
* Will use Joi validator to check:

```For profile```:

- Gender - string, valid(‘Male’, ’Female’, ’Prefer Not say’)
- Date of Birth - date, iso, 
- Gender: string,
- Age: number,
- Nationality: string,
- Preferred Language: string
- Preferred Currency: string
- Department: stringAddress: string
- Line Manager: string
- Phone Number: number
- Marital Status: string, valid(‘Single’, ‘Married’, ‘Widowed’)

#8. Logic on the access (Explaining who is allowed to CRUD (if applicable) the entities you are working on)

* User: Allowed to view and update their profile after registration

#9. Logic on actual Implementation

- Update User model for user 
- Setup the validations using Joi
- Use middleware to check if user is logged in, is profile owner
- Create controllers for profile retrieve and update
- Create the route for profile retrieve and update
- Test the routes with mocha & chai
- Document the endpoint with swagger

#10. Endpoints Format: {url}/api/user/….

GET   - /profile/:id: Get single profile - Owner
PUT   - /profile/:id: update user data - Owner

### CODE REVIEW

File: src/middlewares/profile.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  const tmp = req.files.profileImage.tempFilePath;
+  const Result = await uploader.upload(
+    tmp,
+    { folder: 'Barefoot Nomad' },
+    (_, result) => result
+  );
+  return Result;
+};
+
+export default blogImage;
```
Comment (ericus123) [inline src/middlewares/profile.js]: Why is this in middlewares ????

File: src/services/user.service.js [hunk 2]
```diff
@@ -28,6 +28,28 @@ class UserService {
...
+    return updateduser;
+  }
+
+  static async deleteUser(param) {
+    const user = await User.destroy({
+      where: param,
+    });
+    return user;
+  }
 }
```
Comment (ericus123) [inline src/services/user.service.js]: Is this needed when completing user profile??

File: src/swagger/user.doc.js [hunk 3]
```diff
@@ -79,6 +79,136 @@ const paths = {
...
+        }),
+        500: response('Error occured while fetching profiles', {
+          error: {
+            type: 'string',
+            example: 'Error occured while fetching profiles',
+          },
+        }),
+      },
+    },
+  },
```
Comment (ericus123) [inline src/swagger/user.doc.js]: How are you getting all users????
This doesn't make sense at all.
Your task is complete user profile

**Review by ericus123 — CHANGES_REQUESTED** [review]
Make sure you read your task description carefully and never push again with test codes commented out

File: src/controllers/profile.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        )
+      );
+    }
+  }
+
+  static async updateProfile(req, res, next) {
+    const { id } = req.params;
+    try {
+      const user = await findUser({ id });
+
```
Comment (ericus123) [inline src/controllers/profile.js]: This way is wrong.
You have the id in the token and a user to complete the profile must be 
authenticated first. no need to add an unnecessary param (**id**)

**Review by ericus123 — CHANGES_REQUESTED** [review]

**Review by itisWasp — APPROVED** [review]

File: test/usertest.spec.js [hunk 5]
```diff
@@ -1,163 +1,162 @@
...
+//       expect(res.body).to.be.a('object');
+//       expect(res.body)
+//         .to.have.property('error')
+//         .eql(`${dummyEmailUser.email} is not a valid Email`);
+//     });
+//     after(async () => {
+//       await models.User.destroy({ where: {}, truncate: true });
+//     });
+//   });
+// });
```
Comment (ericus123) [inline test/usertest.spec.js:162]: @Procech20  Why are these tests commented out ???

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: test/usertest.spec.js (hunk 5 above)
Comment (Procech20) [reply test/usertest.spec.js:162]: Working on this ASAP, I thought I uncommented them

**Pull request closed without merging**